lib/bridge/IrcHandler.js

```diff
@@ -9,13 +9,22 @@ var MatrixRoom = require("matrix-appservice-bridge").MatrixRoom;
 var MatrixUser = require("matrix-appservice-bridge").MatrixUser;
 var MatrixAction = require("../models/MatrixAction");
 
+var DISCONNECT_PRESENCE_TIMEOUT = 15 * 60 * 1000; // 15mins
+
 function IrcHandler(ircBridge) {
     this.ircBridge = ircBridge;
     // maintain a map of which user ID is in which PM room, so we know if we
     // need to re-invite them if they bail.
     this._roomIdToPrivateMember = {
         // room_id: { user_id: $USER_ID, membership: "join|invite|leave|etc" }
     };
+    this._rejoinPromises = {
+        // nick: Promise that resolves if the user joins a channel having quit
+        //       previously
+    };
+    this._rejoinResolvers = {
+        // nick: function to call when a user joins
```

Comment:
Keying off the nick alone is not sufficient. It needs to be a `(server.domain, nick)` tuple.